### ISSUE TITLE
Good results even with random initialization

### ISSUE THREAD

**Issue by jonkoi** [danieltan07/dagmm#19]
I managed to get great result when testing consistently even when DAGMM is untrained. The results look similar to this:

Accuracy : 0.9637, Precision : 0.9573, Recall : 0.9312, F-score : 0.9441

I only changed from the cholesky method to direct torch.det for det_cov calculation. The rest is the same as this repository.
For testing random DAGMM, I commented out the code for loading pretrained model and have all pretained models removed so I don't think there was accidental loading. The 4 mus and covs are equal to each other so I believe no training was done.

Do you have an idea where the issue might be?

This seems to contradict many issues posted here, so I am quite baffled.

**Comment by herozen97** [comment]
Hello, I rewrited with torch.det too. @jonkoi 
However, I met a problem.
When I used `(Cholesky.apply(cov_k.cpu() * (2*np.pi)).diag().prod())`, the answer was `tensor(0.0001, grad_fn=<ProdBackward0>)`
When `torch.det(cov_k * (2*np.pi))` was used, the answer was `tensor(1.1221e-08, device='cuda:0', grad_fn=<DetBackward>)`
Why there was a differece?

**Comment by Linguanbiao** [comment]
I also encountered the same problem. The accuracy of testing directly without training is very high, and the total loss has not converged during the training process. What is the matter?

**Comment by DamienFourure** [comment]
Hello. Same problem here. I managed to train it and I had very good results. But then, seeing this issue, I trained it again for 0 epoch and still get 0.94 F1 score.

**Comment by meihuameii** [comment]
@herozen-China  Do you have found what's wrong with `(Cholesky.apply(cov_k.cpu() * (2*np.pi)).diag().prod())` and `torch.det(cov_k * (2*np.pi))`? Thanks a lot.

**Comment by DamienFourure** [comment]
We may have found an answer to this problem. We investigate a lot with my team and we discover two things:

- First: if you randomly project the data to a lower dimensionality (with a randomly initialized fully connected) and apply a basic model (One Class SVM) you also obtain good results. Our hypothesis is that the datasets are actually quite easy and all anomalous samples are far away from the normal distribution (the distribution of normal samples).
- Secondly, and the most important point, is that the F1 score is highly sensitive to the contamination rate of the **test set**. So the way you are splitting your data can change a lot your F1 score. If you split it randomly and then re-inject all the anomalous sample in the test set (like it's done in this code), you'll increase the contamination rate of the test set and improve a lot your F1 score. We actually write a paper about this phenomenon which has just been accepted to the ECML-PKDD 2021 conference. 
The paper is available here: [https://arxiv.org/abs/2106.16020](https://arxiv.org/abs/2106.16020)
And the code to reproduce our results is here: [https://github.com/euranova/F1-Score-is-Biased](https://github.com/euranova/F1-Score-is-Biased)

Do not hesitate to write to me if you have any questions (or feedbacks)

**Comment by MiaoHu-Pro** [comment]
I think the loss function is wrong. That is why the total loss has not converged during the training process.
in the training process, the weights of NNs have not been changed. But the results of the experiments are very good, that is funny...

**Comment by DamienFourure** [comment]
I didn't looked at the loss function since a long time so I cannot tell if it's wrong or not but even if the loss function is wrong, results are very good even without training. The problem is the way the model is evaluated. By putting all anomalous sample like it's done in the code, and estimating the contamination rate on the test set they arbitrally increase the F1-Score. A correct evaluation protocol would be to split randomly into train / test, use the train set to evaluate the contamination rate and then get rid off the anomalous sample in the train set.

**Comment by liuyoude** [comment]
@jonkoi @DamienFourure @Linguanbiao @meihuameii Is the problem of the threshold selection？ The code link: https://github.com/danieltan07/dagmm/blob/0f79614675386a7766a30b9b6a6d1f24e6290528/solver.py#L244. It uses train and test data to select threshold, but I think only the train data can be used to select threshold. This operation caused the good results under random initialization.

**Comment by liuyoude** [comment]
> @jonkoi @DamienFourure @Linguanbiao @meihuameii Is the problem of the threshold selection？ The code link:
> 
> https://github.com/danieltan07/dagmm/blob/0f79614675386a7766a30b9b6a6d1f24e6290528/solver.py#L244
> 
> . It uses train and test data to select threshold, but I think only the train data can be used to select threshold. This operation caused the good results under random initialization.

And then, if the test sample energy is wrong: https://github.com/danieltan07/dagmm/blob/0f79614675386a7766a30b9b6a6d1f24e6290528/solver.py#L231
I think it shoud use train_phi, train_mu and train_cov to calculate test sample energy.
